From a13e27336a61a3452e1ab8b4bb132c4174c67498 Mon Sep 17 00:00:00 2001
From: Michael Niedermayer <michael@niedermayer.cc>
Date: Thu, 15 Oct 2026 01:42:48 +0200
Subject: [PATCH] avcodec/vvc/refs: take the SPS and PPS references only once
 the frame buffer is allocated

Found during triage/review of the security report q0RFgfMenefI
(ANT-2026-HRT03RV0), whose control file vvc-subpic-control-corner.mkv
triggers it.

Fixes: memleak
Regression since: e70225e0a8
Memleak Replicated through UnModified FFmpeg with ASAN

(cherry picked from commit a13e27336a61a3452e1ab8b4bb132c4174c67498)
---
 libavcodec/vvc/refs.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/libavcodec/vvc/refs.c b/libavcodec/vvc/refs.c
index 6939a22732d..c50cd32f5c0 100644
--- a/libavcodec/vvc/refs.c
+++ b/libavcodec/vvc/refs.c
@@ -130,9 +130,6 @@ static VVCFrame *alloc_frame(VVCContext *s, VVCFrameContext *fc)
         if (frame->frame->buf[0])
             continue;
 
-        frame->sps = av_refstruct_ref_c(fc->ps.sps);
-        frame->pps = av_refstruct_ref_c(fc->ps.pps);
-
         // Add LCEVC SEI metadata here, as it's needed in get_buffer()
         if (fc->sei.common.itut_t35.lcevc) {
             ret = ff_frame_new_side_data_from_buf(s->avctx, frame->frame,
@@ -145,6 +142,9 @@ static VVCFrame *alloc_frame(VVCContext *s, VVCFrameContext *fc)
         if (ret < 0)
             return NULL;
 
+        frame->sps = av_refstruct_ref_c(fc->ps.sps);
+        frame->pps = av_refstruct_ref_c(fc->ps.pps);
+
         frame->rpl = av_refstruct_allocz(s->current_frame.nb_units * sizeof(RefPicListTab));
         if (!frame->rpl)
             goto fail;
-- 
2.53.0

